### PR TITLE
[Amazon] Add extractor

### PR DESCRIPTION
**Pull request by Ashish0804**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [ ] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [ ] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Closes https://github.com/yt-dlp/yt-dlp/issues/1509

### CODE REVIEW

File: yt_dlp/extractor/amazon.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        webpage = self._download_webpage(url, id)
+        data_json = self._parse_json(self._html_search_regex(r'var\s?obj\s?=\s?jQuery\.parseJSON\(\'(.*)\'\)', webpage, 'data'), id)
+        entries = []
+        for video in data_json.get('videos') or []:
+            vid_url = video.get('url')
+            if vid_url and video.get('isVideo', False):
+                entries.append({
+                    'id': video['marketPlaceID'],
+                    'title': video.get('title'),
+                    'thumbnail': video.get('thumb') or video.get('thumbUrl'),
```
Comment (u-spec-png) [inline yt_dlp/extractor/amazon.py:53]:
```suggestion
                    'thumbnail': video.get('thumbUrl') or video.get('thumb'),
```

Comment (u-spec-png) [reply yt_dlp/extractor/amazon.py:53]: `thumbUrl` has beater quality so `thumb` shoud be used as fall back.

File: .gitignore [hunk 2]
```diff
@@ -106,3 +106,6 @@ yt-dlp.zip
 ytdlp_plugins/extractor/*
 !ytdlp_plugins/extractor/__init__.py
 !ytdlp_plugins/extractor/sample.py
+
+# Developer Test Generator
+gentest.py
```
Comment (pukkandan) [inline .gitignore:111]:
```suggestion
```

File: yt_dlp/utils.py [hunk 3]
```diff
@@ -4657,7 +4657,7 @@ def parse_codecs(codecs_str):
     vcodec, acodec, hdr = None, None, None
     for full_codec in split_codecs:
         codec = full_codec.split('.')[0]
-        if codec in ('avc1', 'avc2', 'avc3', 'avc4', 'vp9', 'vp8', 'hev1', 'hev2', 'h263', 'h264', 'mp4v', 'hvc1', 'av01', 'theora', 'dvh1', 'dvhe'):
+        if codec in ('avc1', 'avc2', 'avc3', 'avc4', 'vp9', 'vp8', 'hev1', 'hev2', 'h263', 'h264', 'mp4v', 'hvc1', 'av01', 'theora', 'dvh1', 'dvhe', 'vp09'):
```
Comment (pukkandan) [inline yt_dlp/utils.py:4660]:
```suggestion
```

File: yt_dlp/extractor/amazon.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            if vid_url and video.get('isVideo', False):
+                entries.append({
+                    'id': video['marketPlaceID'],
+                    'title': video.get('title'),
+                    'thumbnail': video.get('thumbUrl') or video.get('thumb'),
+                    'duration': video.get('durationSeconds'),
+                    'height': int_or_none(video.get('videoHeight')),
+                    'width': int_or_none(video.get('videoWidth')),
+                    'url': vid_url,
+                })
```
Comment (pukkandan) [inline yt_dlp/extractor/amazon.py:58]:
```suggestion
        entries = [{
            'id': video['marketPlaceID'],
            'url': video['url'],
            'title': video.get('title'),
            'thumbnail': video.get('thumbUrl') or video.get('thumb'),
            'duration': video.get('durationSeconds'),
            'height': int_or_none(video.get('videoHeight')),
            'width': int_or_none(video.get('videoWidth')),
        } for video in data_json.get('videos') or [] if video.get('isVideo') and video.get('url')]
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/amazon.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+from ..utils import int_or_none
+
+
+class AmazonIE(InfoExtractor):
+    _VALID_URL = r'(?:https?://)(?:www\.)?amazon\.(?:[a-z]{2,3})(?:\.[a-z]{2})?/[^/]*/?dp/(?P<id>[^/&#$?]+)'
```
Comment (nyuszika7h) [inline yt_dlp/extractor/amazon.py:9]:
```suggestion
    _VALID_URL = r'(?:https?://)(?:www\.)?amazon\.(?:[a-z]{2,3})(?:\.[a-z]{2})?/[^/]*/?(?:dp|gp/product)/(?P<id>[^/&#$?]+)'
```

**Review by nyuszika7h — CHANGES_REQUESTED** [review]

**Pull request closed without merging**